### ISSUE TITLE
Model convergence problem; non-positive-definite Hessian matrix. 

### ISSUE THREAD

**Issue by statlink** [glmmTMB/glmmTMB#591]
Hello, I have a single reponse variable at 5 distinct years and I fit the following model.

m <- glmmTMB( Y ~ 1 + (1 | id) + ar1(year -1 | id), family = gaussian)

I receive the following

Warning message:
In fitTMB(TMBStruc) :
  Model convergence problem; non-positive-definite Hessian matrix. See vignette('troubleshooting')

> m
Formula:          Y ~ 1 + (1 | id) + ar1(etos - 1 | id)
     AIC      BIC   logLik df.resid 
      NA       NA       NA    17737 
Random-effects (co)variances:

Conditional model:
Error in r[2, 1] : subscript out of bounds

**What is going on?**

**Comment by bbolker** [comment]
Could we have a reproducible example please ... ?

**Comment by statlink** [comment]
I attached my data.

[W1.txt](https://github.com/glmmTMB/glmmTMB/files/4743019/W1.txt)

**Comment by bbolker** [comment]
it helps a lot to use `ar1(factor(year) -1 | id)`.  Obviously we need to do a better job detecting/throwing an error if it's not, and documenting it, and still trying to figure out why the resulting model still has a non-pos-def Hessian ...

**Comment by bbolker** [comment]
@kaskr, any advice on the non-pos-def Hessian troubleshooting below?  I computed the Hessian using `numDeriv::jacobian()` and it seemed OK, but `sdreport()` is reporting problems. (I don't know what's done internally in `sdreport`: `numDeriv::jacobian()` uses Richardson extrapolation so maybe it's more accurate ... ?)  Are there any useful resources/guidelines on debugging the cases where `sdreport()$pdhess` is FALSE ... ?

The (long/unwieldy) `diagnose_hessian()` function included below is also in the "troubleshooting" vignette in [this branch](https://github.com/glmmTMB/glmmTMB/tree/posdef_vignette) ... if we like it, it could be added as a utility in the package?


```
dd <- read.table("W1.txt",header=TRUE)
library(glmmTMB)
m0 <- glmmTMB( Y ~ 1 + (1 | id) + ar1(factor(year) -1 | id), family = gaussian,  data=dd)
diagnose_hessian <- function(fit,h=NULL, eval.eps=1e-5,evec.eps=1e-2) {
    ## pull out the TMB object from the fit
    obj <- fit$obj
    ee <- environment(obj$fn)
    ## extract parameters
    pp <- ee$last.par[-ee$random]
    ## easiest way to get names corresponding to all of the parameters
    nn <- tryCatch(colnames(vcov(fit,full=TRUE)),
                   ## fall-back position
                   error = function(e) make.unique(names(pp)))
    ## fit$sdr$pdHess
    if ("sdr" %in% names(fit)) {
        cat("bad params according to sdreport:",
            paste(nn[!is.finite(suppressWarnings(sqrt(diag(fit$sdr$cov.fixed))))],
                  collapse=", "),"\n")
    }
    ## two ways to compute the Hessian
    ## (1) directly from the objective function, via finite difference+Richardson extrapolation
    ## h1 <- hessian(obj$fn, pp)
    ## (2) use the gradient and compute its Jacobian (faster and probably more stable)
    if (is.null(h)) {
        if (!require(numDeriv)) stop("need numDeriv package installed") 
        h <- jacobian(obj$gr, pp)
    }
    ## double-check we get the same answer (approximately)
    ## all.equal(h1,h,tolerance=1e-5)
    ## now investigate the Hessian
    eigs <- eigen(h)
    ## non-positive definite means some of the eigenvectors are <= 0
    bad <- which(eigs$values/max(eigs$values)<=eval.eps)
    if (length(bad)==0) {
        cat("Hessian seems OK\n")
        return(invisible(h))
    }
    cat(sprintf("max eigenvalue = %1.3g",eigs$values[1]),"\n")
    for (b in bad) {  ## there could be more than one 'bad' direction/eigenvector ..
        cat(sprintf("Hessian eigenvalue %d = %1.3g (relative val = %1.3g)",
                    b,eigs$values[b],eigs$values[b]/eigs$values[1]),"\n")
        bad_vec <- eigs$vectors[,b]
        bad_elements <- which(abs(bad_vec)>evec.eps)
        cat("   bad elements:",nn[bad_elements],"\n")
    }
    cat("SDs computed from sqrt(diag(solve(H))):",
        paste(suppressWarnings(sqrt(diag(solve(h)))), collapse=", "),"\n")
    return(invisible(h))
}
diagnose_hessian(m0)
```

I get:

```
bad params according to sdreport: betad, theta.1, theta.2 
max eigenvalue = 2.78e+04 
Hessian eigenvalue 5 = 0.0936 (relative val = 3.37e-06) 
   bad elements: betad theta.1 
SDs computed from sqrt(diag(solve(H))): 0.0655140891235769, 3.26796470696029, 0.0181964958601156, 0.0850592964228774, 0.0122786323452572 
```

**Comment by statlink** [comment]
I get the same warning when trying with a Hurdle model, family =  trunaced_nbinom2.

**Comment by bbolker** [comment]
@kaskr, bump: any ideas?

**Comment by kaskr** [comment]
A few comments:
@bbolker The `pdHess` flag is set to false by `sdreport` if it fails to cholesky factorize the hessian. In terms of numerical accuracy I don't expect this to be optimal, but it is fast and works most of the time. I agree that for `glmmTMB` it may be worth considering alternative ways to detect PD (at the cost of being slower).
I quickly looked at `diagnose_hessian` (did not test it) and can confirm that `jacobian(obj$gr, pp)` is in general a more accurate approximation of the hessian than that provided by `optimHess`. It is also my impression that `eigen(h)` is more accurate at determining PD than e.g. `chol(h)`. An idea for improvement could be to take a few newton steps from `pp` as the accuracy of the solution might also affect the accuracy of `h`.

**Comment by bbolker** [comment]
I think these issues might be resolved by https://github.com/glmmTMB/glmmTMB/pull/605 ; @statlink, can you try `remotes::install_github("glmmTMB/glmmTMB/glmmTMB@better_pdcheck")` (you'll need compilers installed too) and see if it helps?

@kaskr: I didn't implement your last idea ("take a few newton steps from pp as the accuracy of the solution might also affect the accuracy of h") (yet?) but I'm certainly open to it.

**Comment by mmaechler** [comment]
I strongly agree with @kaskr 's proposal to ask for higher convergence accuracy in the case one wants to use the approximate Hessian (provided by the optimizer which needs/uses that often anyway) for inference (vcov(.) !) or even just checking pos.def.

**Comment by bbolker** [comment]
@kaskr, can you explain your idea a little bit more? Is the idea that the nonlinear optimizer might not have converged to the solution sufficiently well? And that this would only be done in the case where `pdHess` is FALSE?  Would we use `h` as computed by `jacobian()` for the first reported optim and iterate without updating it (which I learned is called a "dishonest Newton method": https://scicomp.stackexchange.com/questions/30202/difference-between-dishonest-newton-method-and-very-dishonest-newton-method ), or ... ?
